test-apps/appui-test-app/appui-test-providers/src/ui/dialogs/ISVPIPView.tsx

```diff
@@ -0,0 +1,188 @@
+
+import "./ISVPIPView.scss";
+import { Id64String } from "@itwin/core-bentley";
+import ViewDefinitionSelector, { getViewDefinitions } from "../components/ViewDefinitionSelector";
+import { ViewIdChangedEventArgs, ViewportComponentEvents } from "@itwin/imodel-components-react";
+import { Presentation, TRANSIENT_ELEMENT_CLASSNAME } from "@itwin/presentation-frontend";
+import { KeySet } from "@itwin/presentation-common";
+interface ISVPIPViewDefInterfaceLocal {
+  id: string; class: string; label: string;
+}
+export function ISVPIPView({ contentId, showViewPicker }: { contentId: string, showViewPicker?: boolean }) {
+  const getIds=(args: Readonly<KeySet>) =>{
+    let allIds: Set<string> = new Set<string>();
+    args.instanceKeys.forEach((ids: Set<string>, key: string) => {
+    // Avoid transient elements
+      if (key !== TRANSIENT_ELEMENT_CLASSNAME) allIds = new Set([...allIds, ...ids]);
+    });
+    return allIds;
+  };
+  const menuItems: AbstractMenuItemProps[] = React.useMemo(() => {
+    return [
+      {
+        id: "Item1", label: "Item ~1", icon: "icon-placeholder",
+        submenu: [
+          { id: "0", item: { label: "SubMenu Item ~1", icon: "icon-placeholder", execute: () => { } } },
+          { id: "1", item: { label: "SubMenu Item ~2", icon: "icon-placeholder", execute: () => { } } },
+        ],
+      },
+      {
+        id: "Item2", item: { label: "Item ~2", icon: "icon-placeholder", execute: () => { } },
+      },
+      {
+        id: "Item3", item: { label: "Item ~3", icon: "icon-placeholder", execute: () => { } },
+      },
+    ];
+  }, []);
+
+  const activeIModelConnection = useActiveIModelConnection();
+  const divRef = React.useRef<HTMLDivElement>(null);
+
+  const [initialViewState, setInitialViewState] = React.useState<ViewState | undefined>(undefined);
+  const [twoDViewDefinitions, settwoDViewDefinitions] = React.useState<ISVPIPViewDefInterfaceLocal[]>([]);
+  const [threeDViewDefinitions, setthreeDViewDefinitions] = React.useState<ISVPIPViewDefInterfaceLocal[]>([]);
+
+  const handleViewIdChange = React.useCallback (async (args: ViewIdChangedEventArgs) => {
+    if (args.newId === args.oldId)
+      return;
+
+    const ids = getIds(Presentation.selection.getSelection(UiFramework.getIModelConnection()!, 0));
+    if ([...ids].length === 0) {
+      ViewManip.fitView(args.viewport as ScreenViewport, false);
+    }
+    const activeFrontstageDef = FrontstageManager.activeFrontstageDef;
+
+    const viewContentControl = activeFrontstageDef?.contentControls?.find((contentControl) => contentControl.viewport === args.viewport);
+
+    if (!viewContentControl)
+      return;
+
+    const mainViewportOnFrontstage = activeFrontstageDef?.contentControls[0].viewport;
+    // turn off listener while we update the views
+    ViewportComponentEvents.onViewIdChangedEvent.removeListener(handleViewIdChange);
```

Comment:
Since there is no `await` and we are not putting the listener back in a `then`, and all potentially time consuming opreations between the removal and addnig back are promises (` xxx.view.load`) the listener will be put back before the loading process really happen, which mostly seems to defeat the reason for removing it in the first place.

Reply:
You are correct. This was only properly called when the callback was executed when no changes were required (e.g., one of the views changed but not to a new dimension.) I've changed how the listener is restored.